File: Baekjoon/2252번: 줄세우기.py
from collections import deque
import sys

n, m = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(n+1)]
visited = [False] * (n+1)
in_degree = [0] * (n+1)
for _ in range(m):
  a, b = map(int, sys.stdin.readline().split())
  graph[a].append(b)
  in_degree[b] += 1

answer = []
def topological_sort():
  q = deque()
  for i in range(1, n+1):
    if in_degree[i] == 0:
      q.append(i)
      visited[i] = True
  while q:
    v = q.popleft()
    answer.append(v)
    for i in graph[v]:
      if not(visited[i]):
        in_degree[i] -= 1
        if in_degree[i] == 0:
          q.append(i)
          visited[i] = True
v = []
topological_sort()
print(*answer, sep=' ')
        

File: Baekjoon/2252번: 줄 세우기.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int n, m;
vector<int> graph[32001];
vector<int> in_degree(32001, 0);
vector<bool> visited(32001, false);
queue<int> q;
void topological_sort(){
  for(int i=1; i < n+1; i++){
    if(in_degree[i] == 0){
      q.push(i);
    }
  }

  while(!(q.empty())){
    int v = q.front();
    q.pop();
    cout << v << ' ';
    for(int i: graph[v]){
      if (!visited[i]){
        in_degree[i]--;
        if(in_degree[i] == 0){
          q.push(i);
          visited[i] = true;
          }
        }
      }
    }
  }
int main() {
  cin >> n >> m;
  
  for(int i; i < m; i++){
    int a, b;
    cin >> a >> b;
    graph[a].emplace_back(b);
    in_degree[b]++;
  }
  topological_sort();
    return 0;
}#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int n, m;
vector<int> graph[32001];
vector<int> in_degree(32001, 0);
vector<bool> visited(32001, false);
queue<int> q;
void topological_sort(){
  for(int i=1; i < n+1; i++){
    if(in_degree[i] == 0){
      q.push(i);
    }
  }

  while(!(q.empty())){
    int v = q.front();
    q.pop();
    cout << v << ' ';
    for(int i: graph[v]){
      if (!visited[i]){
        in_degree[i]--;
        if(in_degree[i] == 0){
          q.push(i);
          visited[i] = true;
          }
        }
      }
    }
  }
int main() {
  cin >> n >> m;
  
  for(int i; i < m; i++){
    int a, b;
    cin >> a >> b;
    graph[a].emplace_back(b);
    in_degree[b]++;
  }
  topological_sort();
    return 0;
}
